### PR TITLE
Extracted theming into own bundle

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | See UPGRADE.md
| Fixed tickets | none
| Related issues/PRs | https://github.com/sulu/SuluThemeBundle/pull/1 https://github.com/sulu/sulu-standard/pull/703
| License | MIT
| Documentation PR | https://github.com/sulu/sulu-docs/pull/225

#### What's in this PR?

Extracted liip-theme bundle.

#### Why?

We want to make sulu-usage easy for simple websites.

##### Webspaces

We have deprecated (1.0) the schema for webspaces and created a new version (1.1) of it. 

```
<?xml version="1.0" encoding="utf-8"?>
<webspace xmlns="http://schemas.sulu.io/webspace/webspace"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://schemas.sulu.io/webspace/webspace http://schemas.sulu.io/webspace/webspace-1.1.xsd">
          
          ...
          
</webspace>
```

You should update your webspace.xml files soonish. To do that you simply have to move the `default-templates` and `error-templates` from the `theme` node and put it into the `webspace` node after the `theme`.
 
The theme is now optional and can be used with a theme-bundle. Sulu has extracted this functionality to make it replaceable with any theming bundle you want. To keep the old directory-structure and functionality please read the next part of this file.

##### Theming

If you have multiple themes (or you don't want to change the folder structure of you project) you have to include the bundle https://github.com/sulu/SuluThemeBundle in your abstract kernel.

This bundle contains all the code which is necessary to use theming in your application.

#### To Do

- [x] Create a documentation PR
- [x] Move error- and default-template from theme to webspace
- [x] Document BC-Breaks
- [x] Move theme-bundle to https://github.com/sulu/SuluThemeBundle

### CODE REVIEW

**Comment by danrot** [comment]
Shouldn't we also move the `EventAwareEngine` to the new bundle? Its sole purpose is to set the theme early enough. Maybe we could even include that logic in a `ThemeAwareEngine`, instead of using the `EventDispatcher` (at least if we can't think of any usecase where we need this via events).

**Comment by wachterjohannes** [comment]
i will think of this after @webmozart has reviewed this "roughly".

**Comment by wachterjohannes** [comment]
i have also moved the eventaware engine

File: src/Sulu/Bundle/CategoryBundle/Tests/app/Resources/webspaces/sulu.io.xml [hunk 1]
```diff
@@ -14,9 +14,6 @@
 
     <theme>
         <key>default</key>
```
Comment (chirimoya) [inline src/Sulu/Bundle/CategoryBundle/Tests/app/Resources/webspaces/sulu.io.xml]: `<theme>default</theme>`

File: src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php [hunk 2]
```diff
@@ -39,13 +38,8 @@ public function postGenerateAction()
...
         $parts = $this->getRequestParameter($this->getRequest(), 'parts', true);
         $templateKey = $this->getRequestParameter($this->getRequest(), 'template', true);
-
-        list($webspaceKey, $languageCode) = $this->getWebspaceAndLanguage();
-        if ($templateKey === null) {
-            $webspaceManager = $this->container->get('sulu_core.webspace.webspace_manager');
-            $webspace = $webspaceManager->findWebspaceByKey($webspaceKey);
-            $templateKey = $webspace->getTheme()->getDefaultTemplate(Structure::TYPE_PAGE);
-        }
+        $webspaceKey = $this->getRequestParameter($this->getRequest(), 'webspace', true);
```
Comment (chirimoya) [inline src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php]: ???

File: src/Sulu/Bundle/MediaBundle/DependencyInjection/ImageFormatCompilerPass.php [hunk 3]
```diff
@@ -34,10 +34,15 @@ class ImageFormatCompilerPass implements CompilerPassInterface
     public function process(ContainerBuilder $container)
     {
         $this->container = $container;
-        $container->setParameter(
-            'sulu_media.image.formats',
-            $this->loadThemeFormats($container->getParameter('sulu_media.format_manager.default_imagine_options'))
+
+        $formats = $this->loadThemeFormats(
```
Comment (chirimoya) [inline src/Sulu/Bundle/MediaBundle/DependencyInjection/ImageFormatCompilerPass.php]: `$this->loadImageFormats()`

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php [hunk 4]
```diff
@@ -91,14 +92,26 @@ public function supports($resource, $type = null)
      */
     private function parseXml($file)
     {
+        $exception = null;
+
         // load xml file
         try {
-            $xmlDoc = XmlUtils::loadFile($file, __DIR__ . static::SCHEME_PATH);
+            $xmlDoc = XmlUtils::loadFile($file, __DIR__ . static::SCHEME_PATH_1_1);
```
Comment (chirimoya) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: use loop for validation

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php [hunk 5]
```diff
@@ -129,6 +145,16 @@ private function parseXml($file)
         return $this->webspace;
     }
 
+    private function tryLoad($file, $scheme)
```
Comment (chirimoya) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: ???

File: src/Sulu/Component/Webspace/Loader/schema/webspace/webspace-1.1.xsd [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        </xs:sequence>
+    </xs:complexType>
+    <xs:complexType name="resource-locatorType">
+        <xs:sequence>
+            <xs:element type="xs:string" name="strategy"/>
+        </xs:sequence>
+    </xs:complexType>
+    <xs:complexType name="themeType">
+        <xs:sequence>
+            <xs:element type="xs:string" name="key"/>
```
Comment (chirimoya) [inline src/Sulu/Component/Webspace/Loader/schema/webspace/webspace-1.1.xsd]: remove key

File: src/Sulu/Component/Webspace/Theme.php [hunk 7]
```diff
@@ -23,34 +23,11 @@ class Theme implements ArrayableInterface
     private $key;
```
Comment (chirimoya) [inline src/Sulu/Component/Webspace/Theme.php]: remove class

File: src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php (hunk 2 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php]: the template key is mandatory and the function `getWebspaceAndLanguage` does this two function calls.

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader.php (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader.php]: wasnt called before i have removed it.

**Comment by chirimoya** [comment]
ACK

**Pull request merged**